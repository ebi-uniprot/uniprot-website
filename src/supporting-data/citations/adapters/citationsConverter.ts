import { SetRequired } from 'type-fest';

import { Statistics } from '../../../shared/types/apiModel';
import { Evidence } from '../../../uniprotkb/types/modelTypes';

export enum CitationXRefDB {
  PubMed = 'PubMed',
  DOI = 'DOI',
}

export type CitationXRefDBType = `${CitationXRefDB}`;

export type CitationXRef = {
  database?: CitationXRefDBType | string;
  id?: string;
};

export type CitationType =
  | 'journal article'
  | 'book'
  | 'online journal article'
  | 'patent'
  | 'submission'
  | 'thesis'
  | 'UniProt indexed literatures'
  | 'unpublished observations';

export type Citation = {
  // Either a pubmed ID, or a CI-<hash> internal hash if pubmed unavailable, or
  // IND<...> also available in the data
  id: `${number}` | `CI-${string}` | `IND${string}`;
  citationType?: CitationType;
  authors?: string[];
  citationCrossReferences?: CitationXRef[];
  title?: string;
  // Will always be a string, could be like '2000', or 'DEC-2000'.
  publicationDate?: string;
  institute?: string;
  patentNumber?: string;
  bookName?: string;
  editors?: string[];
  journal?: string;
  locator?: string;
  firstPage?: string;
  lastPage?: string;
  volume?: string;
  publisher?: string;
  address?: string;
  completeAuthorList?: boolean;
  literatureAbstract?: string;
  authoringGroup?: string[];
  submissionDatabase?: string;
};

export type SourceCategory =
  | 'Function'
  | 'Names'
  | 'Subcellular Location'
  | 'Disease & Variants'
  | 'Phenotypes & Variants'
  | 'PTM / Processing'
  | 'Expression'
  | 'Interaction'
  | 'Structure'
  | 'Family & Domains'
  | 'Sequences';

export type ReferenceComment = {
  value: string;
  type: string;
  evidences?: Evidence[];
};

export type Reference = {
  citationId?: Citation['id'];
  referencePositions?: string[];
  referenceComments?: ReferenceComment[];
  evidences?: Evidence[];
  source?: { name: string; id?: string };
  pubMedId?: string;
  sourceCategories?: SourceCategory[];
  referenceNumber?: number; // Only for UniProtKB (trembl and swissprot)
  communityAnnotation?: CommunityAnnotation; // Only for community annotations
  annotation?: string; // Only for computationally mapped
};

export interface CommunityAnnotation {
  proteinOrGene?: string;
  function?: string;
  comment?: string;
  disease?: string;
}

export type CitationsAPIModel = {
  statistics?: SetRequired<
    Statistics,
    'computationallyMappedProteinCount' | 'communityMappedProteinCount'
  > & {
    largeScale?: boolean;
  };
  citation: Citation;
  references?: Reference[];
};

export type CitationsUIModel = CitationsAPIModel & {
  // any addition/change by the converter
};

const getCitationPubMedId = (citation: Citation) =>
  citation.citationCrossReferences?.find((xref) => xref.database === 'PubMed');

const getDoiXref = (citation: Citation) =>
  citation.citationCrossReferences?.find((xref) => xref.database === 'DOI');

// Note, should this be done as part of citationsConverter?
export const formatCitationData = (citation: Citation) => {
  const pubMedXref = getCitationPubMedId(citation);

  const doiXref = getDoiXref(citation);

  const pubmedId = pubMedXref && pubMedXref.id;

  const journalInfo = {
    journal: citation.journal,
    volume: citation.volume,
    firstPage: citation.firstPage,
    lastPage: citation.lastPage,
    publicationDate: citation.publicationDate,
    doiId: doiXref?.id,
    submissionDatabase: citation?.submissionDatabase,
    citationType: citation?.citationType,
    locator: citation?.locator,
    editors: citation.editors,
    bookName: citation.bookName,
    publisher: citation.publisher,
    address: citation.address,
    patentNumber: citation.patentNumber,
  };

  return { pubmedId, journalInfo };
};

const citationsConverter = (data: CitationsAPIModel): CitationsUIModel => ({
  ...data,
});

export default citationsConverter;
