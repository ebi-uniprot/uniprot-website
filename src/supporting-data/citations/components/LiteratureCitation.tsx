import { createElement, useState, FC, Fragment, HTMLAttributes } from 'react';
import { Link } from 'react-router-dom';
import cn from 'classnames';
import {
  Bubble,
  Button,
  PublicationIcon,
  ComputerMappedIcon,
  CommunityAnnotationIcon,
  SwissProtIcon,
  TremblIcon,
  EllipsisReveal,
} from 'franklin-sites';
import { capitalize } from 'lodash-es';
import { SetOptional } from 'type-fest';

import ExternalLink from '../../../shared/components/ExternalLink';

import {
  getEntryPath,
  Location,
  LocationToPath,
} from '../../../app/config/urls';
import externalUrls from '../../../shared/config/externalUrls';

import parseDate from '../../../shared/utils/parseDate';
import cleanText, {
  cleanTextDefaultOptions,
  getTransformTags,
} from '../../../shared/utils/cleanText';

import {
  CitationsAPIModel,
  CitationType,
  formatCitationData,
} from '../adapters/citationsConverter';
import { Namespace } from '../../../shared/types/namespaces';

import '../../../shared/styles/literature-citation.scss';

type AuthorProps = {
  authors?: string[];
  authoringGroup?: string[];
  limit?: number;
};

const getLinkToAuthor = (author: string) => ({
  pathname: LocationToPath[Location.UniProtKBResults],
  search: `query=lit_author:"${author}"`,
});

const WORM_BREEDERS_GAZETTE_URL = 'http://www.wormbook.org/wli/';
// const ARVO_URL = 'http://iovs.arvojournals.org/article.aspx?articleid=';
const PGR_URL = 'https://www.ebi.ac.uk/~textman/pgr-htdocs/pgr/';

export const getLocatorUrl = (
  locator?: string,
  name?: 'Plant Gene Register' | "Worm Breeder's Gazette" | string
) => {
  if (!locator || !name) {
    return null;
  }
  switch (name) {
    case 'Plant Gene Register':
      return `${PGR_URL}${locator}.html`;
    case "Worm Breeder's Gazette": {
      const regex = /(\d+)\((\d+)\):(\d+)/;
      const matches = locator?.match(regex);
      if (!matches) {
        return null;
      }
      const issue = matches[1];
      const volume = matches[2];
      const page = matches[3];
      return `${WORM_BREEDERS_GAZETTE_URL}wbg${issue}.${volume}p${page}/`;
    }
    default:
      return null;
  }
};

const getChoppedAuthorLists = (authors: string[], limit: number) => {
  const cutoff = Math.max(authors.length, limit);

  const displayedAuthors = authors.slice(0, limit - 1);
  const hiddenAuthors = authors.slice(limit - 1, cutoff - 1);
  const lastAuthor = authors.slice(cutoff - 1);
  return { displayedAuthors, hiddenAuthors, lastAuthor };
};

const Authors: FC<AuthorProps> = ({ authors, authoringGroup, limit = 10 }) => {
  const { displayedAuthors, hiddenAuthors, lastAuthor } = getChoppedAuthorLists(
    authors || [],
    limit
  );

  return (
    <div className="publication__authors">
      {authoringGroup?.map((group, index) => (
        // eslint-disable-next-line react/no-array-index-key
        <Fragment key={index}>
          {index !== 0 && ', '}
          <Link to={getLinkToAuthor(group)}>{group}</Link>
        </Fragment>
      ))}
      {displayedAuthors.map((author, index) => (
        // eslint-disable-next-line react/no-array-index-key
        <Fragment key={index}>
          {index !== 0 && ', '}
          <Link to={getLinkToAuthor(author)}>{author}</Link>
        </Fragment>
      ))}
      {hiddenAuthors.length > 0 && (
        <EllipsisReveal>
          {hiddenAuthors.map((author, index) => (
            // eslint-disable-next-line react/no-array-index-key
            <Fragment key={index}>
              , <Link to={getLinkToAuthor(author)}>{author}</Link>
            </Fragment>
          ))}
        </EllipsisReveal>
      )}
      {lastAuthor.length === 1 && (
        <>
          {', '}
          <Link to={getLinkToAuthor(lastAuthor[0])}>{lastAuthor[0]}</Link>
        </>
      )}
    </div>
  );
};

type AbstractProps = {
  abstract: string;
  open?: boolean;
};

const Abstract: FC<AbstractProps> = ({ abstract, open = false }) => {
  const [display, setDisplay] = useState(open);
  return (
    <div className="publication__abstract">
      <Button variant="tertiary" onClick={() => setDisplay(!display)}>
        {display ? 'Hide' : 'View'} abstract
      </Button>
      {display && (
        <p
          // eslint-disable-next-line react/no-danger
          dangerouslySetInnerHTML={{
            __html: cleanText(abstract, {
              ...cleanTextDefaultOptions,
              transformTags: {
                ...getTransformTags('strong'),
              },
            }),
          }}
        />
      )}
    </div>
  );
};

type JournalInfoProps = {
  journalInfo: {
    publicationDate?: string;
    journal?: string;
    firstPage?: string;
    lastPage?: string;
    volume?: string;
    doiId?: string;
    submissionDatabase?: string;
    citationType?: CitationType;
    locator?: string;
    editors?: string[];
    bookName?: string;
    publisher?: string;
    address?: string;
    patentNumber?: string;
  };
};

export const JournalInfo: FC<JournalInfoProps> = ({
  journalInfo: {
    publicationDate,
    journal,
    firstPage,
    lastPage,
    volume,
    doiId,
    submissionDatabase,
    citationType,
    locator,
    editors,
    bookName,
    publisher,
    address,
    patentNumber,
  },
}) => {
  let name = journal || doiId;

  if (citationType === 'book' && bookName) {
    name =
      editors !== undefined ? `${editors?.join(',')}; ${bookName}` : bookName;
  }

  let page = null;
  if (firstPage) {
    page = firstPage;
    if (lastPage) {
      page += `-${lastPage} `;
    }
  }

  let date;
  if (publicationDate) {
    date = (
      <>
        (
        <time dateTime={parseDate(publicationDate)?.toISOString()}>
          {publicationDate}
        </time>
        )
      </>
    );
  }

  const noDisplayCitationTypes: CitationType[] = [
    'UniProt indexed literatures',
    'journal article',
  ];

  const content = (
    <>
      {citationType && !noDisplayCitationTypes.includes(citationType) && (
        <div className="tiny">{capitalize(citationType)}</div>
      )}
      {name} {volume}
      {volume && page && ':'}
      {page}
      {submissionDatabase && `Submitted to ${submissionDatabase} `}
      {patentNumber && `Patent publication ${patentNumber} `}
      {publisher}
      {publisher && address && ', '}
      {address}
      {date}
    </>
  );

  const url = doiId ? externalUrls.DOI(doiId) : getLocatorUrl(locator, name);

  if (!url) {
    return content;
  }

  return <ExternalLink url={url}>{content}</ExternalLink>;
};

type StatisticsProps = {
  statistics: CitationsAPIModel['statistics'];
  id: number | string;
};

const Statistics: FC<StatisticsProps> = ({ statistics, id }) => {
  const {
    reviewedProteinCount,
    unreviewedProteinCount,
    computationallyMappedProteinCount,
    communityMappedProteinCount,
  } = statistics || {};
  return (
    <div className="publication__statistics">
      {reviewedProteinCount || unreviewedProteinCount ? (
        <div className="publication__statistics__content">
          <small>Cited in</small>
          <div className="publication__statistics__content__body">
            {reviewedProteinCount ? (
              <Link
                to={{
                  pathname: LocationToPath[Location.UniProtKBResults],
                  search: `facets=reviewed:true&query=(lit_citation_id:${id})`,
                }}
                title={`UniProtKB reviewed entries: ${reviewedProteinCount}`}
              >
                <Bubble size="small">{reviewedProteinCount}</Bubble>
                <SwissProtIcon
                  className="publication__statistics__content__body--reviewed"
                  height="1em"
                />
              </Link>
            ) : undefined}
            {unreviewedProteinCount ? (
              <Link
                to={{
                  pathname: LocationToPath[Location.UniProtKBResults],
                  search: `facets=reviewed:false&query=(lit_citation_id:${id})`,
                }}
                title={`UniProtKB unreviewed entries: ${unreviewedProteinCount}`}
              >
                <Bubble size="small">{unreviewedProteinCount}</Bubble>
                <TremblIcon
                  className="publication__statistics__content__body--unreviewed"
                  height="1em"
                />
              </Link>
            ) : undefined}
          </div>
        </div>
      ) : undefined}
      {computationallyMappedProteinCount || communityMappedProteinCount ? (
        <div className="publication__statistics__content">
          <small>Mapped to</small>
          <div className="publication__statistics__content__body">
            {computationallyMappedProteinCount ? (
              <Link
                to={{
                  pathname: LocationToPath[Location.UniProtKBResults],
                  // NOTE: only works for PubMed IDs
                  search: `query=(computational_pubmed_id:${id})`,
                }}
                title={`Computationally mapped entries: ${computationallyMappedProteinCount}`}
              >
                <Bubble size="small">
                  {computationallyMappedProteinCount}
                </Bubble>
                <ComputerMappedIcon height="1em" />
              </Link>
            ) : undefined}
            {communityMappedProteinCount ? (
              <Link
                to={{
                  pathname: LocationToPath[Location.UniProtKBResults],
                  // NOTE: only works for PubMed IDs
                  search: `query=(community_pubmed_id:${id})`,
                }}
                title={`Community mapped entries: ${communityMappedProteinCount}`}
              >
                <Bubble size="small">{communityMappedProteinCount}</Bubble>
                <CommunityAnnotationIcon height="1em" />
              </Link>
            ) : undefined}
          </div>
        </div>
      ) : undefined}
    </div>
  );
};

const LiteratureCitation: FC<
  {
    data: SetOptional<CitationsAPIModel, 'statistics'>;
    displayAll?: boolean;
    headingLevel?: `h${1 | 2 | 3 | 4 | 5 | 6}`;
    linkToEntry?: boolean;
  } & HTMLAttributes<HTMLElement>
> = ({
  data,
  displayAll,
  headingLevel = 'h5',
  children,
  linkToEntry,
  className,
  ...props
}) => {
  const { citation, statistics } = data;
  const { title, authors, literatureAbstract, authoringGroup } = citation;
  const { pubmedId, journalInfo } = formatCitationData(citation);

  const headingContent = title || <em>No title available.</em>;

  const heading = createElement(
    headingLevel,
    {
      className: cn('publication__heading', 'small', {
        'publication__heading--no-title': !title,
      }),
    },
    linkToEntry ? (
      <Link to={getEntryPath(Namespace.citations, citation.id)}>
        {headingContent}
      </Link>
    ) : (
      headingContent
    )
  );

  return (
    <article className={cn(className, 'publication')} {...props}>
      <div className="publication__columns">
        <div className="publication__columns__main">
          {heading}
          <Authors
            authors={authors}
            authoringGroup={authoringGroup}
            limit={displayAll ? +Infinity : 10}
          />
          {literatureAbstract && (
            <Abstract abstract={literatureAbstract} open={displayAll} />
          )}
          {children}
        </div>
        <div className="publication__columns__side">
          <div className="publication__columns__side__item">
            {pubmedId && <PublicationIcon width="1.875em" height="2em" />}
            <ul className="no-bullet">
              {pubmedId && (
                <>
                  <li>
                    <ExternalLink url={externalUrls.PubMed(pubmedId)}>
                      PubMed
                    </ExternalLink>
                  </li>
                  <li>
                    <ExternalLink url={externalUrls.EuropePMC(pubmedId)}>
                      Europe PMC
                    </ExternalLink>
                  </li>
                </>
              )}
              {journalInfo && (
                <li>
                  <small>
                    <JournalInfo journalInfo={journalInfo} />
                  </small>
                </li>
              )}
            </ul>
          </div>
          {statistics && (
            <div className="publication__columns__side__item">
              <Statistics statistics={statistics} id={citation.id} />
            </div>
          )}
        </div>
      </div>
    </article>
  );
};

export default LiteratureCitation;
