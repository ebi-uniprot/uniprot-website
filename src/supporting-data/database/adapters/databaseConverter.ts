import { Statistics } from '../../../shared/types/apiModel';

export type DatabaseAPIModel = {
  pubMedId: string;
  abbrev: string;
  doiId: string;
  linkType: string;
  dbUrl: string;
  category: string;
  name: string;
  id: string;
  servers: string[];
  statistics?: Statistics;
};

export type DatabaseUIModel = DatabaseAPIModel & {
  // any addition/change by the converter
};

const databaseConverter = (data: DatabaseAPIModel): DatabaseUIModel => ({
  ...data,
});

export default databaseConverter;
