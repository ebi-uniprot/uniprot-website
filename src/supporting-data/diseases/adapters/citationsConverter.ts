export type CitationsAPIModel = {};

export type CitationsUIModel = CitationsAPIModel & {
  // any addition/change by the converter
};

const citationsConverter = (data: CitationsAPIModel): CitationsUIModel => ({
  ...data,
});

export default citationsConverter;
