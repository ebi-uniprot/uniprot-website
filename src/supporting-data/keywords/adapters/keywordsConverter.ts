import { Statistics } from '../../../shared/types/apiModel';

type GO = {
  name: string;
  // Swagger says "id"
  // id: string;
  // But endpoint gives
  goId: `GO:${string}`;
};

export type KeywordNameID = {
  name: string;
  id: `KW-${string}`;
};

export type KeywordsLite = {
  keyword: KeywordNameID;
  links?: string[];
  definition: string;
  synonyms?: string[];
  geneOntologies?: GO[];
  category?: KeywordNameID;
  children?: KeywordsLite[];
};

export type KeywordsAPIModel = KeywordsLite & {
  accession?: string;
  statistics?: Statistics;
  parents?: KeywordsLite[];
};

export type KeywordsUIModel = KeywordsAPIModel & {
  // any addition/change by the converter
};

const keywordsConverter = (data: KeywordsAPIModel): KeywordsUIModel => ({
  ...data,
});

export default keywordsConverter;
