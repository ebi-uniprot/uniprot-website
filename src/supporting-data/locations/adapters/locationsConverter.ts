type GO = {
  name: string;
  id: string;
};

type Statistics = {
  reviewedProteinCount: number;
  unreviewedProteinCount: number;
};

type Keyword = {
  name: string;
  id: string;
};

export type LocationsAPIModel = {
  references: string[];
  geneOntologies: GO[];
  note: string;
  isA: null[]; // TODO: review
  partOf: null[]; // TODO: review
  statistics: Statistics;
  synonyms: string[];
  // TODO: change to enum of possible values, example: 'LOCATION'
  // TODO: review, maybe it's always 'LOCATION'...
  category: string;
  name: string;
  id: string;
  definition: string;
  content: string;
  keyword: Keyword;
  links: string[];
};

export type LocationsUIModel = LocationsAPIModel & {
  // any addition/change by the converter
};

const locationsConverter = (data: LocationsAPIModel): LocationsUIModel => ({
  ...data,
});

export default locationsConverter;
