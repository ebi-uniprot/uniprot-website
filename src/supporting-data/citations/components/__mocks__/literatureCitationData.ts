import { CitationsAPIModel } from '../../adapters/citationsConverter';

// uniprot/api/citations/14702039
const mock: CitationsAPIModel = {
  citation: {
    citationType: 'UniProt indexed literatures',
    authors: [
      'Ota T.',
      'Suzuki Y.',
      'Nishikawa T.',
      'Otsuki T.',
      'Sugiyama T.',
      'Irie R.',
      'Wakamatsu A.',
      'Hayashi K.',
      'Sato H.',
      'Nagai K.',
      'Kimura K.',
      'Makita H.',
      'Sekine M.',
      'Obayashi M.',
      'Nishi T.',
      'Shibahara T.',
      'Tanaka T.',
      'Ishii S.',
      'Yamamoto J.',
      'Saito K.',
      'Kawai Y.',
      'Isono Y.',
      'Nakamura Y.',
      'Nagahari K.',
      'Murakami K.',
      'Yasuda T.',
      'Iwayanagi T.',
      'Wagatsuma M.',
      'Shiratori A.',
      'Sudo H.',
      'Hosoiri T.',
      'Kaku Y.',
      'Kodaira H.',
      'Kondo H.',
      'Sugawara M.',
      'Takahashi M.',
      'Kanda K.',
      'Yokoi T.',
      'Furuya T.',
      'Kikkawa E.',
      'Omura Y.',
      'Abe K.',
      'Kamihara K.',
      'Katsuta N.',
      'Sato K.',
      'Tanikawa M.',
      'Yamazaki M.',
      'Ninomiya K.',
      'Ishibashi T.',
      'Yamashita H.',
      'Murakawa K.',
      'Fujimori K.',
      'Tanai H.',
      'Kimata M.',
      'Watanabe M.',
      'Hiraoka S.',
      'Chiba Y.',
      'Ishida S.',
      'Ono Y.',
      'Takiguchi S.',
      'Watanabe S.',
      'Yosida M.',
      'Hotuta T.',
      'Kusano J.',
      'Kanehori K.',
      'Takahashi-Fujii A.',
      'Hara H.',
      'Tanase T.-O.',
      'Nomura Y.',
      'Togiya S.',
      'Komai F.',
      'Hara R.',
      'Takeuchi K.',
      'Arita M.',
      'Imose N.',
      'Musashino K.',
      'Yuuki H.',
      'Oshima A.',
      'Sasaki N.',
      'Aotsuka S.',
      'Yoshikawa Y.',
      'Matsunawa H.',
      'Ichihara T.',
      'Shiohata N.',
      'Sano S.',
      'Moriya S.',
      'Momiyama H.',
      'Satoh N.',
      'Takami S.',
      'Terashima Y.',
      'Suzuki O.',
      'Nakagawa S.',
      'Senoh A.',
      'Mizoguchi H.',
      'Goto Y.',
      'Shimizu F.',
      'Wakebe H.',
      'Hishigaki H.',
      'Watanabe T.',
      'Sugiyama A.',
      'Takemoto M.',
      'Kawakami B.',
      'Yamazaki M.',
      'Watanabe K.',
      'Kumagai A.',
      'Itakura S.',
      'Fukuzumi Y.',
      'Fujimori Y.',
      'Komiyama M.',
      'Tashiro H.',
      'Tanigami A.',
      'Fujiwara T.',
      'Ono T.',
      'Yamada K.',
      'Fujii Y.',
      'Ozaki K.',
      'Hirao M.',
      'Ohmori Y.',
      'Kawabata A.',
      'Hikiji T.',
      'Kobatake N.',
      'Inagaki H.',
      'Ikema Y.',
      'Okamoto S.',
      'Okitani R.',
      'Kawakami T.',
      'Noguchi S.',
      'Itoh T.',
      'Shigeta K.',
      'Senba T.',
      'Matsumura K.',
      'Nakajima Y.',
      'Mizuno T.',
      'Morinaga M.',
      'Sasaki M.',
      'Togashi T.',
      'Oyama M.',
      'Hata H.',
      'Watanabe M.',
      'Komatsu T.',
      'Mizushima-Sugano J.',
      'Satoh T.',
      'Shirai Y.',
      'Takahashi Y.',
      'Nakagawa K.',
      'Okumura K.',
      'Nagase T.',
      'Nomura N.',
      'Kikuchi H.',
      'Masuho Y.',
      'Yamashita R.',
      'Nakai K.',
      'Yada T.',
      'Nakamura Y.',
      'Ohara O.',
      'Isogai T.',
      'Sugano S.',
    ],
    citationCrossReferences: [
      {
        database: 'PubMed',
        id: '14702039',
      },
      {
        database: 'DOI',
        id: '10.1038/ng1285',
      },
    ],
    title:
      'Complete sequencing and characterization of 21,243 full-length human cDNAs.',
    publicationDate: '2004',
    journal: 'Nat. Genet.',
    firstPage: '40',
    lastPage: '45',
    volume: '36',
    completeAuthorList: true,
    literatureAbstract:
      'As a base for human transcriptome and functional genomics, we created the "full- length long Japan" (FLJ) collection of sequenced human cDNAs. We determined the entire sequence of 21,243 selected clones and found that 14,490 cDNAs (10,897 clusters) were unique to the FLJ collection. About half of them (5,416) seemed to be protein-coding. Of those, 1,999 clusters had not been predicted by computational methods. The distribution of GC content of nonpredicted cDNAs had a peak at approximately 58% compared with a peak at approximately 42%for predicted cDNAs. Thus, there seems to be a slight bias against GC-rich transcripts in current gene prediction procedures. The rest of the cDNAs unique to the FLJ collection (5,481) contained no obvious open reading frames (ORFs) and thus are candidate noncoding RNAs. About one-fourth of them (1,378) showed a clear pattern of splicing. The distribution of GC content of noncoding cDNAs was narrow and had a peak at approximately 42%, relatively low compared with that of protein-coding cDNAs.',
  },
  references: [
    {
      source: {
        name: 'UniProtKB reviewed (Swiss-Prot)',
      },
      pubMedId: '14702039',
      sourceCategories: ['Sequence'],
      referenceComments: [
        {
          value: 'Kidney',
          type: 'TISSUE',
        },
        {
          value: 'Thalamus',
          type: 'TISSUE',
        },
      ],
      referencePositions: [
        'NUCLEOTIDE SEQUENCE [LARGE SCALE MRNA] (ISOFORM 2)',
      ],
      referenceNumber: 1,
    },
  ],
  statistics: {
    reviewedProteinCount: 10309,
    unreviewedProteinCount: 2615,
    computationallyMappedProteinCount: 0,
    communityMappedProteinCount: 0,
  },
};

export default mock;
