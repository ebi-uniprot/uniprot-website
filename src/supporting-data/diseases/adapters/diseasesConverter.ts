import { Statistics } from '../../../shared/types/apiModel';
import { KeywordNameID } from '../../keywords/adapters/keywordsConverter';

type XRef = {
  databaseType: string;
  properties?: string[];
  id: string;
};

export type DiseasesAPIModel = {
  keywords?: KeywordNameID[];
  alternativeNames?: string[];
  acronym: string;
  crossReferences: XRef[];
  name: string;
  id: string;
  definition: string;
  statistics?: Statistics;
};

export type DiseasesUIModel = DiseasesAPIModel & {
  // any addition/change by the converter
};

const diseasesConverter = (data: DiseasesAPIModel): DiseasesUIModel => ({
  ...data,
});

export default diseasesConverter;
