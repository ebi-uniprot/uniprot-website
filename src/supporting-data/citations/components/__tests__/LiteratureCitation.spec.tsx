import { fireEvent, screen } from '@testing-library/react';

import customRender from '../../../../shared/__test-helpers__/customRender';
import { CitationsAPIModel } from '../../adapters/citationsConverter';
import literatureCitationData from '../__mocks__/literatureCitationData';
import LiteratureCitation, { getLocatorUrl } from '../LiteratureCitation';

let rendered: ReturnType<typeof customRender>;

describe('Publication component', () => {
  beforeEach(() => {
    rendered = customRender(
      <LiteratureCitation data={literatureCitationData['14702039']} />
    );
  });

  it('should render', () => {
    const { asFragment } = rendered;
    expect(asFragment()).toMatchSnapshot();
  });

  it('should expand authors', async () => {
    expect(screen.queryByText('Ohara O.')).not.toBeInTheDocument();
    fireEvent.click(screen.getByRole('button', { name: '[...]' }));
    expect(screen.getByText('Ohara O.')).toBeInTheDocument();
  });

  it('should display 1 author', async () => {
    const mockData: CitationsAPIModel = {
      citation: {
        id: '14702039',
        authors: ['Smith X.'],
      },
    };
    rendered.rerender(<LiteratureCitation data={mockData} />);
    expect(screen.getByText('Smith X.')).toBeInTheDocument();
  });

  it('should expand abstract', async () => {
    expect(screen.queryByText(/noncoding cDNAs/)).not.toBeInTheDocument();
    fireEvent.click(screen.getByRole('button', { name: 'View abstract' }));
    expect(screen.getByText(/noncoding cDNAs/)).toBeInTheDocument();
  });

  test('Publication component with submissionDatabase if present', () => {
    rendered.rerender(
      <LiteratureCitation data={literatureCitationData['CI-5GBDQ6B103N1E']} />
    );
    expect(screen.getByText(/Submitted to/)).toBeInTheDocument();
  });
});

describe('Publication component url', () => {
  test('Plant Gene Register', () => {
    const url = getLocatorUrl('PGR98-023', 'Plant Gene Register');
    expect(url).toEqual(
      'https://www.ebi.ac.uk/~textman/pgr-htdocs/pgr/PGR98-023.html'
    );
  });

  test("Worm Breeder's Gazette", () => {
    const url = getLocatorUrl('14(2):46', "Worm Breeder's Gazette");
    expect(url).toEqual('http://www.wormbook.org/wli/wbg14.2p46/');
  });

  test("Wrong locator for Worm Breeder's", () => {
    const url = getLocatorUrl('', "Worm Breeder's Gazette");
    expect(url).toBeNull();
  });
});
