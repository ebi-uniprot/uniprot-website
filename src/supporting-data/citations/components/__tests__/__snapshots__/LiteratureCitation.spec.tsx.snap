// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Publication component should render 1`] = `
<DocumentFragment>
  <article
    class="publication"
  >
    <div
      class="publication__columns"
    >
      <div
        class="publication__columns__main"
      >
        <h5
          class="publication__heading tiny"
        >
          Complete sequencing and characterization of 21,243 full-length human cDNAs.
        </h5>
        <div
          class="publication__authors"
        >
          <a
            href="/uniprotkb?query=lit_author:\\"Ota T.\\""
          >
            Ota T.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Suzuki Y.\\""
          >
            Suzuki Y.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Nishikawa T.\\""
          >
            Nishikawa T.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Otsuki T.\\""
          >
            Otsuki T.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Sugiyama T.\\""
          >
            Sugiyama T.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Irie R.\\""
          >
            Irie R.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Wakamatsu A.\\""
          >
            Wakamatsu A.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Hayashi K.\\""
          >
            Hayashi K.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Sato H.\\""
          >
            Sato H.
          </a>
          <button
            aria-expanded="false"
            class="button tertiary ellipsis-reveal"
            title="Show more"
            type="button"
          >
            [...]
          </button>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Sugano S.\\""
          >
            Sugano S.
          </a>
        </div>
        <div
          class="publication__abstract"
        >
          <button
            class="button tertiary"
            type="button"
          >
            View abstract
          </button>
        </div>
      </div>
      <div
        class="publication__columns__side"
      >
        <div
          class="publication__columns__side__item"
        >
          <test-file-stub
            height="2em"
            width="1.875em"
          />
          <ul
            class="no-bullet"
          >
            <li>
              <a
                class="external-link"
                href="https://pubmed.ncbi.nlm.nih.gov/14702039"
                rel="noopener"
                target="_blank"
              >
                PubMed
              </a>
            </li>
            <li>
              <a
                class="external-link"
                href="//europepmc.org/article/MED/14702039"
                rel="noopener"
                target="_blank"
              >
                Europe PMC
              </a>
            </li>
            <li>
              <small>
                <a
                  class="external-link"
                  href="https://dx.doi.org/10.1038/ng1285"
                  rel="noopener"
                  target="_blank"
                >
                  Nat. Genet. 36:40-45(
                  <time
                    datetime="2004-01-01T00:00:00.000Z"
                  >
                    2004
                  </time>
                  )
                  <test-file-stub
                    data-testid="external-link-icon"
                    width="12.5"
                  />
                </a>
              </small>
            </li>
          </ul>
        </div>
        <div
          class="publication__columns__side__item"
        >
          <small>
            Mapped to
          </small>
          <div
            class="publication__statistics"
          >
            <a
              href="/uniprotkb?facets=reviewed:true&query=(lit_citation_id:14702039)"
              title="UniProtKB reviewed entries: 32172"
            >
              <span
                class="bubble--small colour-pastel-blue"
              >
                99+
              </span>
              <test-file-stub
                classname="publication__statistics--reviewed"
                height="1em"
              />
            </a>
            <a
              href="/uniprotkb?facets=reviewed:false&query=(lit_citation_id:14702039)"
              title="UniProtKB unreviewed entries: 2615"
            >
              <span
                class="bubble--small colour-pastel-blue"
              >
                99+
              </span>
              <test-file-stub
                classname="publication__statistics--unreviewed"
                height="1em"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
